gen_data/gen_test_m24_single.py

```diff
@@ -0,0 +1,475 @@
+        sampled_vids.append(np_videos[sample_idx])
+        if current_sample_idx >= num_times_to_sample:
+            break
+        current_sample_idx += 1
+    sampled_vids = np.array(sampled_vids)
+    assert sampled_vids.shape == (num_times_to_sample, num_samples)
+    return sampled_vids
+
+def get_known_and_unknown(train, test, novelty_type):
+    if novelty_type == 'relation' or novelty_type == 'relation_type':
+        known_classes = set(train[novelty_type + '1'].dropna()).union(
+            set(train[novelty_type + '2'].dropna()))
+        unknown_classes = set(test[novelty_type + '1'].dropna()).union(
+            (test[novelty_type + '2'].dropna()))
+        unknown_classes -= known_classes
+    else:
+        assert novelty_type in ['class','perspective', 'ontology_id','location']
+        known_classes = set(train[novelty_type].dropna())
+        unknown_classes = set(test[novelty_type].dropna()) - known_classes
+    assert len(known_classes) > 0
+    assert len(unknown_classes) > 0
+    known_classes = [x.strip().lower() if isinstance(x, str) else x for x in list(known_classes)]
+    unknown_classes = [x.strip().lower() if isinstance(x, str) else x for x in list(unknown_classes)]
+    return_dict_known, return_dict_unknown = OrderedDict(), OrderedDict()
+    for i in range(test.shape[0]):
+        row = test.iloc[i]
+        label = row['ontology_id']
+        vid_id = row['anonymous_id']
+        assert vid_id not in return_dict_unknown
+        assert vid_id not in return_dict_known
+        
+        if novelty_type == 'relation' or novelty_type == 'relation_type':
+            is_novel = False
+            for count in ['1','2']:
+                novelty_attr = row[novelty_type + count]
+                if isinstance(novelty_attr,str):
+                    novelty_attr = novelty_attr.strip().lower()
+                elif np.isnan(novelty_attr):
+                    continue
+                if novelty_attr in known_classes:
+                    pass 
+                elif novelty_attr in unknown_classes:
+                    is_novel = True
+                else:
+                    raise ValueError(f'Novelty attr {novelty_attr} isn\'t in known or unknown')
+
+            if is_novel:
+                return_dict_unknown[vid_id] = label
+            else:
+                return_dict_known[vid_id] = label
+        else:
+            novelty_attr = row[novelty_type] 
+            if isinstance(novelty_attr,str):
+                novelty_attr = novelty_attr.strip().lower()
+            elif np.isnan(novelty_attr):
+                continue
+            if novelty_attr in known_classes:
+                return_dict_known[vid_id] = label
+            elif novelty_attr in unknown_classes:
+                return_dict_unknown[vid_id] = label
+            else:
+                raise ValueError(f'Novelty attr {novelty_attr} isn\'t in known or unknown')
+
+    return return_dict_known, return_dict_unknown
+
+def gen_test(
```

Comment:
Should this be `gen_tests`? Since it looks like it is generating a bunch of tests